### DISCUSSION TITLE
Caching issue when generating files in the build folder

### DISCUSSION THREAD

**Discussion by MeganAlexia** [webpack/webpack#16875] [Q&A]
# Bug report

**What is the current behavior?**
This is a minimal reproduction of the issue we are facing - we are trying to generate files inside the build folder at build time thanks to a custom loader and a custom plugin. However, this has been behaving randomly and the files aren't always moved and populated properly, which means that the different pages can't access the files at runtime. We have managed to narrow it done to the cache as every time we delete the build folder, it would be working. Using `useCache: false` in the webpack config means the issue isn't happening.

**If the current behavior is a bug, please provide the steps to reproduce.**
Minimal reproduction on a NextJS app: https://github.com/MeganAlexia/webpack-nextjs-bug-repro 
* Run `yarn` to install the dependencies.
* Make sure to delete the /build folder if you have one.
* Run yarn dev and open "localhost:3000", it should be an empty page with "Hello world".  
* We are expecting the custom plugin (https://github.com/MeganAlexia/webpack-nextjs-bug-repro/blob/main/webpack/plugin/test-plugin.js) and the custom loader (https://github.com/MeganAlexia/webpack-nextjs-bug-repro/blob/main/webpack/loader/test-loader.js) to have run. In the IDE, open /build/testing/text.txt and make sure it's populated with the string ", one, two, three, four".
* Kill the yarn dev and re-run it. Re-check /build/testing/text.txt - it should be an empty file. /!\ You might need to do this step a couple of times as the issue appears randomly and not necessarily on the second run.

**What is the expected behavior?**
Every time `yarn dev` is running, the file "text.txt" should be created in the `/build/testing` folder and should be populated with the relevant string (in this case, ", one, two, three, four" for example).

**Other relevant information:**
webpack version: 5.76.2
Node.js version: 18.15.0
- Operating System: 
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
- Additional tools:
      next: 13.2.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      Yarn: 1.22.19

**Comment by alexander-akait** [comment]
This code is wrong https://github.com/MeganAlexia/webpack-nextjs-bug-repro/blob/main/webpack/plugin/test-plugin.js#L31:
- You should not use `compiler.hooks.emit` to emit assets, we have `compiler.hooks.compilation` and ` compilation.hooks.processAssets`
- You should not write assets manually using real fs, you need to emit them

Take a look at the simple JSON minimizer plugin https://github.com/webpack-contrib/json-minimizer-webpack-plugin/blob/master/src/index.js:
- For you case you need https://github.com/webpack-contrib/json-minimizer-webpack-plugin/blob/master/src/index.js#L199 to use another stage - `PROCESS_ASSETS_STAGE_ADDITIONAL` - https://github.com/webpack/webpack/blob/main/lib/Compilation.js#L5223
 - For loader you can use `emitFile` - https://webpack.js.org/api/loaders/#thisemitfile, if you need to emit something like assets
 - Don't forget to cache it to avoid extra execution (you can find an example of logic in `json-minimizer-webpack-plugin`)
 - To emit assets in a plugin you can use https://webpack.js.org/api/compilation-object/#emitasset
 
 Feel free to feedback

**Reply by MeganAlexia** [reply]
Hi!  

 First of all, thanks for the detailed explanations and moving this to discussion. This is really helpful! 
 
 I had a look at what you have sent over, and I think my use-case might be a little different. I have a folder at the root of the directory containing some files. At build time, I would like to be able to grab these files in the Webpack compiler environment and generate assets from them that will then be moved into the build folder.   In the examples you have sent over (and what I’ve found on the internet in general), operations seems to always be performed on assets. I was wondering if there is any way for me to grab files that are not considered assets yet and read them? If not, what would be the recommended way to go? 
 
 Apologies, this sounds like quite a newbie question but I couldn't find what I was looking for in the documentation...
 
   Thanks a lot,  
   
 Megan

**Reply by alexander-akait** [reply]
>  I was wondering if there is any way for me to grab files that are not considered assets yet and read them?

Can you clarify? Because, for example, async `import(...)` creates a chunk which will be JS assets... For example [it is an example fo](https://github.com/yoriiis/svg-chunk-webpack-plugin/) is plugin to create sprites and read assets
